src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala

```diff
@@ -0,0 +1,252 @@
+
+  private def copyFirstFile(df: DataFrame): Unit = {
+    val files = df.queryExecution.optimizedPlan.collect {
+      case LogicalRelation(
+          HadoopFsRelation(location: PartitioningAwareFileIndex, _, _, _, _, _),
+          _,
+          _,
+          _) =>
+        location.allFiles
+    }.flatten
+    val sourcePath = files.head.getPath
+    val destPath = new Path(files.head.getPath + ".copy")
+    sourcePath.getFileSystem(new Configuration).copyToLocalFile(sourcePath, destPath)
+  }
+
+  override def beforeAll(): Unit = {
+    super.beforeAll()
+    val sparkSession = spark
+    import sparkSession.implicits._
+    hyperspace = new Hyperspace(spark)
+    FileUtils.delete(new Path(sampleParquetDataLocation))
+    FileUtils.delete(new Path(sampleJsonDataLocation))
+    val dfFromSample = sampleData.toDF("Date", "RGUID", "Query", "imprs", "clicks")
+    dfFromSample.write.parquet(sampleParquetDataLocation)
+    dfFromSample.write.json(sampleJsonDataLocation)
+
+    df = spark.read.parquet(sampleParquetDataLocation)
+    hyperspace.createIndex(df, indexConfig1)
+    copyFirstFile(df)
+
+    df = spark.read.json(sampleJsonDataLocation)
+    hyperspace.createIndex(df, indexConfig2)
+    copyFirstFile(df)
+  }
+
+  after {
+    spark.disableHyperspace()
+  }
+
+  override def afterAll(): Unit = {
+    super.afterAll()
+    FileUtils.delete(new Path(sampleParquetDataLocation))
+    FileUtils.delete(new Path(sampleJsonDataLocation))
+  }
+
+  test("HybridScan filter rule test") {
+    df = spark.read.parquet(sampleParquetDataLocation)
+    val query = df.filter(df("clicks") <= 2000).select(df("query"))
+    val expectedResult = query.collect
+
+    withSQLConf("spark.hyperspace.index.hybridscan.enabled" -> "false") {
+      val transformed = FilterIndexRule(query.queryExecution.optimizedPlan)
+      assert(transformed.equals(query.queryExecution.optimizedPlan), "No plan transformation.")
+    }
+
+    withSQLConf("spark.hyperspace.index.hybridscan.enabled" -> "true") {
+      val planWithHybridScan = FilterIndexRule(query.queryExecution.optimizedPlan)
+      assert(
+        !planWithHybridScan.equals(query.queryExecution.optimizedPlan),
+        "Plan should be transformed.")
+
+      // check appended file is added to relation node or not
+      val nodes = planWithHybridScan collect {
+        case p @ LogicalRelation(fsRelation: HadoopFsRelation, _, _, _) =>
+          assert(fsRelation.location.inputFiles.count(_.contains(".copy")) === 1)
+          assert(fsRelation.location.inputFiles.count(_.contains("index1")) === 4)
+          p
+      }
+      assert(nodes.length === 1)
+
+      spark.enableHyperspace()
+      val query2 = df.filter(df("clicks") <= 2000).select(df("query"))
+      val res = query2.collect()
+
+      assert(expectedResult.nonEmpty)
+      assert(res.nonEmpty)
+      assert(expectedResult.sortBy(_.toString).toSeq === res.sortBy(_.toString).toSeq)
```

Comment:
Can't you just do `checkAnswer(query2, expectedResult)` now that you are extending `QueryTest`?